FWCore/SOA/interface/Column.h

```diff
@@ -85,4 +85,9 @@ namespace edm {
   struct _ClassName_ : public edm::soa::Column<_Type_, _ClassName_> { \
     static constexpr const char* const kLabel = _String_;             \
   }
+#define SOA_DECLARE_DEFAULT(_ClassName_, _Expression_)                \
+  template <class Object>                                             \
+  _ClassName_::type value_for_column(Object const& x, _ClassName_*) { \
+    return x._Expression_;                                            \
+  }
```

Comment:
One downside of this approach comes up with a case where for some property some class hierarchy would have an expression that differs from the "default" expression. In such a case each deriving class of the hierarchy would need its own overload of the `value_for_column()` function. An example
```cpp
namespace edm::soa {
  namespace col {
    SOA_DECLARE_COLUMN(Eta, float, "eta");
  }
  SOA_DECLARE_DEFAULT(col::Eta, eta());
}

// base class
class SomeBase {
public:
  double Eta() const; // note upper case
};

// some deriving classes
class SomeDeriving : public SomeBase {...};
class AnotherDeriving : public SomeBase {...};

// override for the base class
double value_for_column(SomeBase const& x, edm::soa::col::Eta *) {
  return x.Eta();
}

// without override for deriving class, the template is chosen over the overload for the base class
// leading to compilation failure
SomeDeriving y;
value_for_column(y, static_cast<edm::soa::col::Eta *>(nullptr);

// therefore, need an overload for each deriving class
double value_for_column(SomeDeriving const& x, edm::soa::col::Eta *) {
  return x.Eta();
}
double value_for_column(AnotherDeriving const& x, edm::soa::col::Eta *) {
  return x.Eta();
}
```

Without a templated "default" an overload for the base class would be sufficient to cover all possible deriving classes as well.

Reply:
Hi @makortel, thanks for providing this sound argument against the macro and taking the time to provide an example! I did not think about this indeed.

Here is the other side of the coin. Fair enough, with this very generic template fillers you have more verbose code in the case you describe, but on the other hand you would have to redefine `value_for_column` for each class that uses the column, even if they have the same interface. For example, not all classes that implement `pt()` or `eta()` have a common base class (see e.g. PR https://github.com/cms-sw/cmssw/pull/30992, where the `etaPhiTable` is applied to a collection of `PFRecTrack` which only inherits from `PFTrack`). Also, since CMSSW is a monorepo, we can always easily change the class such that it implements the "default" interface that the column expects in the same PR (again, PR https://github.com/cms-sw/cmssw/pull/30992 is an example). This might be even seen as a good thing, because it makes interfaces converge. Therefore, I think this interface-driven or "duck type" approach of `SOA_DECLARE_DEFAULT` is not so bad.

Reply:
And in the case where the overload resolution could be a problem, we could work around it by having this macro to define a different function, e.g. `value_for_column_default()` and in the call site we'd call that only if no suitable `value_for_column()` overload is found.

I'd be ready to accept this macro. Could you add a comment describing the caveat above (with the possible workaround in case we ever hit it), and the recommendation to use the macro only in the global namespace?

Could you also extend the unit test to cover this new macro?